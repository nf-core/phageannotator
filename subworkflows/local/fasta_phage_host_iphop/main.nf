//
// Predict phage hosts using iPHoP
//

include { IPHOP_DOWNLOAD    } from '../../../modules/nf-core/iphop/download/main'
include { UNTAR             } from '../../../modules/nf-core/untar/main'
include { UNTAR as UNTAR2   } from '../../../modules/nf-core/untar/main'
include { IPHOP_PREDICT     } from '../../../modules/nf-core/iphop/predict/main'

workflow FASTA_PHAGE_HOST_IPHOP {
    take:
    virus_fasta     // [ [ meta ], fasta ]  , virus sequences (mandatory)
    iphop_db        // [ checkv_db ]        , iPHoP database directory (optional)

    main:
    ch_versions = Channel.empty()

    // if iphop_db exists, skip IPHOP_DOWNLOAD
    if ( iphop_db ){
        ch_iphop_db = iphop_db
    } else {
        // if partial run is specified the minimal, incomplete database will be downloaded
        if ( !params.iphop_partial_test ) {
            //
            // MODULE: download standard IPHoP database
            //
            ch_iphop_db = IPHOP_DOWNLOAD( ).iphop_db
            ch_versions = ch_versions.mix(IPHOP_DOWNLOAD.out.versions)
        } else {
            //
            // MODULE: unpack minimal iphop db
            //
            ch_iphop_db = UNTAR ( [ [ id:'iphop_minimal_db' ], file("https://github.com/nf-core/test-datasets/raw/phageannotator/modules/nfcore/iphop/download/Test_db_rw.tar.gz", checkIfExists: true) ] ).untar.map { it[1] }
            ch_versions = ch_versions.mix(UNTAR.out.versions)
        }
    }

    // create input for partial iphop run
    ch_iphop_partial_input = []
    if ( params.iphop_partial_test ) {
        //
        // MODULE: unpack data for partial iphop run
        //
        ch_iphop_partial_input = UNTAR2 ( [ [ id:'iphop_partial_input' ], file("https://github.com/nf-core/test-datasets/raw/phageannotator/modules/nfcore/iphop/predict/iPHoP_data.tar.gz", checkIfExists: true) ] ).untar.map { it[1] }
        ch_versions = ch_versions.mix(UNTAR2.out.versions)
    }

    //
    // MODULE: Predict virus host
    //
    ch_host_predictions_tsv = IPHOP_PREDICT ( virus_fasta, ch_iphop_db, ch_iphop_partial_input ).iphop_genus
    ch_versions = ch_versions.mix(IPHOP_PREDICT.out.versions.first())

    emit:
    host_predictions_tsv    = ch_host_predictions_tsv   // [ [ meta ], genus_predictions.tsv ]   , TSV file containing host genus predictions
    versions                = ch_versions               // [ versions.yml ]
}
