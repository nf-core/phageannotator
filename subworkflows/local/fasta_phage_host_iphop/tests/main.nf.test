nextflow_workflow {

    name "Test Subworkflow: FASTA_PHAGE_HOST_IPHOP"
    script "../main.nf"
    workflow "FASTA_PHAGE_HOST_IPHOP"

    // Dependencies
    tag "IPHOP_DOWNLOAD"
    tag "UNTAR"
    tag "IPHOP_PREDICT"

    test("fasta") {

        when {
            params {
                iphop_test_db       = true
                iphop_partial_test  = true
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ],
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                    ]
                )
                input[1] = null
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
