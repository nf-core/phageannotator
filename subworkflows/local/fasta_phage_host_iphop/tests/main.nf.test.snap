{
    "fasta": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "Host_prediction_to_genus_m90.csv:md5,528dcefb6a6c743b43f0dbc40bec3692"
                    ]
                ],
                "1": [
                    "versions.yml:md5,41c2135a3578a7a52a23fd3243c22e80",
                    "versions.yml:md5,6d962b190899f9580909b28917ffa748",
                    "versions.yml:md5,c8e1743aa09ef3013ef7ddb1edc8c43e"
                ],
                "host_predictions_tsv": [
                    [
                        {
                            "id": "test"
                        },
                        "Host_prediction_to_genus_m90.csv:md5,528dcefb6a6c743b43f0dbc40bec3692"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,41c2135a3578a7a52a23fd3243c22e80",
                    "versions.yml:md5,6d962b190899f9580909b28917ffa748",
                    "versions.yml:md5,c8e1743aa09ef3013ef7ddb1edc8c43e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-19T17:08:56.56407858"
    }
}
